apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicyElements.cs

```diff
@@ -0,0 +1,459 @@
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Globalization;
+using System.Net;
+
+namespace Google.Cloud.Storage.V1
+{
+    public sealed partial class UrlSigner
+    {
+        /// <summary>
+        /// Represents an element that can be included on a <see cref="PostPolicy"/>.
+        /// </summary>
+        internal interface IPostPolicyElement
+        {
+            /// <summary>
+            /// The name with which the element is included on a <see cref="PostPolicy"/>.
+            /// </summary>
+            string ElementName { get; }
+        }
+
+        /// <summary>
+        /// Represents a <see cref="IPostPolicyElement"/> that supports the exact match condition operation.
+        /// </summary>
+        internal interface ISupportsExact<TValue> : IPostPolicyElement
+        {
+            /// <summary>
+            /// A way to convert the values of <typeparamref name="TValue"/> to
+            /// the representation with which to include them on the <see cref="PostPolicy"/>.
+            /// </summary>
+            /// <param name="value">The value for which to return the corresponding representation.</param>
+            /// <returns>The representation of the given value.</returns>
+            object ToPostPolicyValue(TValue value);
+        }
+
+        /// <summary>
+        /// Represents a <see cref="IPostPolicyElement"/> that supports the range match condition operation.
+        /// </summary>
+        internal interface ISupportsRange<TValue> : IPostPolicyElement
+            where TValue : IComparable
+        {
+            /// <summary>
+            /// A way to convert the values of <typeparamref name="TValue"/> to
+            /// the representation with which to include them on the <see cref="PostPolicy"/>.
+            /// </summary>
+            /// <param name="value">The value for which to return the corresponding representation.</param>
+            /// <returns>The representation of the given value.</returns>
+            object ToPostPolicyValue(TValue value);
+        }
+
+        /// <summary>
+        /// Represents a <see cref="IPostPolicyElement"/> that supports the starts with condition operation.
+        /// </summary>
+        internal interface ISupportsStartsWith : IPostPolicyElement { }
+
+        /// <summary>
+        /// The ACL that you want to allow for the objects that will be uploaded.
+        /// </summary>
+        internal sealed class Acl : ISupportsExact<string>, ISupportsStartsWith
+        {
+            /// <summary>
+            /// The default acl element.
+            /// </summary>
+            public static Acl Element { get; } = new Acl();
+
+            internal Acl() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "acl";
+
+            /// <inheritdoc/>
+            object ISupportsExact<string>.ToPostPolicyValue(string value) => value;
+        }
+
+        /// <summary>
```

Comment:
I suspect with one class for "exact element" and one derived class for "exact and starts with", using private/protected constructors for the element name, you could remove a lot of duplication.

Very happy to experiment with that later though. That's the nice thing about all of this being internal.